tests/app_client/automation.py

```diff
@@ -0,0 +1,182 @@
+        ["button", Button.RIGHT.value, False],
+    ]
+
+    default_rule = {
+        "actions": [
+            ["setbool", "default_match", True]
+        ]
+    }
+
+    @classmethod
+    def actions(cls, acts):
+        return reduce(operator.add, acts)
+
+    @classmethod
+    def rule(
+        cls,
+        actions,
+        text: None,
+        is_regex: bool = False,
+        conditions: Tuple[str, ConditionFlag] = None,
+    ) -> dict:
+        rule = {}
+        if text:
+            if is_regex:
+                rule['regexp'] = text
+            else:
+                rule['text'] = text
+        if conditions:
+            conds = []
+            for varname, flag in conditions:
+                if flag == ConditionFlag.RUN_FOR:
+                    conds.append([varname, True])
+                elif flag == ConditionFlag.NOT_RUN:
+                    conds.append([varname, False])
+                elif flag == ConditionFlag.RUN_ONCE:
+                    conds.append([varname, False])
+                    actions.append(["setbool", varname, True])
+            rule["conditions"] = conds
+        rule["actions"] = actions
+        return rule
+
+    @classmethod
+    def get_address_rule(cls):
+        return [
+            cls.rule(
+                cls.go_left,
+                text='Address',
+            ),
+            cls.rule(
+                cls.press_both,
+                text='Approve',
+            ),
+            cls.default_rule,
+        ]
+
+    @classmethod
+    def get_xpub_rule(cls, confirm: bool):
+        if confirm:
+            acts = cls.actions([cls.go_left, cls.go_left, cls.press_both])
+        else:
+            acts = cls.actions([cls.go_left, cls.press_both])
+        return [
+            cls.rule(
+                acts,
+                text='access?',
+            ),
+            cls.default_rule,
+        ]
+
+    @classmethod
+    def sign_tx_reject_output_rule(cls):
+        acts = cls.actions([cls.go_left, cls.press_both])
+        return [
+            cls.rule(acts, text='Output'),
+            cls.default_rule,
+        ]
+
+    @classmethod
+    def sign_tx_reject_send_rule(cls):
+        output_acts = cls.actions([cls.go_left, cls.go_left, cls.press_both])
+        send_acts = cls.actions([cls.go_left, cls.press_both])
+        return [
+            cls.rule(output_acts, text='Output'),
+            cls.rule(send_acts, text='Transaction?'),
+            cls.default_rule,
+        ]
+
+    @classmethod
+    def sign_tx_accept_rule(cls):
+        output_acts = cls.actions([cls.go_left, cls.go_left, cls.press_both])
+        send_acts = cls.actions([cls.go_left, cls.go_left, cls.press_both])
+        return [
+            cls.rule(output_acts, text='Output'),
+            cls.rule(send_acts, text='Transaction?'),
+            cls.default_rule,
+        ]
+
+    # sign_tx_quit_rule?
+
+
+class Automation(metaclass=ABCMeta):
+    @abstractmethod
+    def set_accept_all(self):
+        ...
```

Comment:
What's this?

Reply:
The Ellipsis operator in this context works as a placeholder for the abstract method code much like a `pass`, this was the style choice of the boilerplate.